src/io/RELION5/RELION5ParticleData.py

```diff
@@ -0,0 +1,205 @@
+
+    def rot2_from_matrix(self, matrix):
+        """rlnAngleTilt -- Theta"""
+        abs_sb = self._abs_sb(matrix)
+
+        if abs_sb is not None:
+            sign_sb = self._sign_rot2(matrix)
+            angle = np.arctan2(sign_sb * abs_sb, matrix[2, 2])
+        else:
+            if np.sign(matrix[2, 2]) > 0:
+                angle = 0
+            else:
+                angle = np.pi
+
+        return angle * 180.0 / np.pi
+
+    def rot3_from_matrix(self, matrix):
+        """Psi"""
+        abs_sb = self._abs_sb(matrix)
+
+        if abs_sb is not None:
+            angle = np.arctan2(matrix[1, 2], -matrix[0, 2])
+        else:
+            if np.sign(matrix[2, 2]) > 0:
+                angle = np.arctan2(-matrix[1, 0], matrix[0, 0])
+            else:
+                angle = np.arctan2(matrix[1, 0], -matrix[0, 0])
+
+        return angle * 180.0 / np.pi
+
+    def _abs_sb(self, matrix):
+        abs_sb = np.sqrt(matrix[0, 2] * matrix[0, 2] + matrix[1, 2] * matrix[1, 2])
+
+        if abs_sb > EPSILON16:
+            return abs_sb
+        else:
+            return None
+
+    def _sign_rot2(self, matrix):
+        rot3 = np.arctan2(matrix[1, 2], -matrix[0, 2])
+
+        if np.abs(np.sin(rot3)) < EPSILON:
+            sign_sb = np.sign(-matrix[0, 2] / np.cos(rot3))
+        else:
+            sign_sb = np.sign(matrix[1, 2]) if (np.sin(rot3) > 0) else -np.sign(matrix[1, 2])
+
+        return sign_sb
+
+
+
+
+class RELION5ParticleData(ParticleData):
+
+
+    DATA_KEYS = {
+        'rlnTomoName': [],
+        'rlnCenteredCoordinateXAngst': [],
+        'rlnCenteredCoordinateYAngst': [],
+        'rlnCenteredCoordinateZAngst': [],
+        'rlnTomoSubtomogramRot': [],
+        'rlnTomoSubtomogramTilt': [],
+        'rlnTomoSubtomogramPsi': [],
+        'rlnAngleRot': [],
+        'rlnAngleTilt': [],
+        'rlnAnglePsi': [],
+        'rlnAngleTiltPrior': [],
+        'rlnAnglePsiPrior': []
+    }
+
+
+    DEFAULT_PARAMS = {
+      'pos_x': 'rlnCenteredCoordinateXAngst',
+      'pos_y': 'rlnCenteredCoordinateYAngst',
+      'pos_z': 'rlnCenteredCoordinateZAngst',
+      'shift_x': 'rlnOriginX',
+      'shift_y': 'rlnOriginY',
+      'shift_z': 'rlnOriginZ',
+      'ang_1': 'rlnAngleRot',
+      'ang_2': 'rlnAngleTilt',
+      'ang_3': 'rlnAnglePsi'
+
+    }
+
+    ROT = RELIONEulerRotation
+
+    def __init__(self, session, file_name, oripix=1, trapix=1, additional_files=None):
+        self.remaining_loops = {}
+        self.remaining_data = {}
+        self.loop_name = 0
+        self.name_prefix = None
+        self.name_leading_zeros = None
+
+        super().__init__(session, file_name, oripix=oripix, trapix=trapix, additional_files=additional_files)
+
+        self.oripix = oripix
+
+
+    #reading of Relion5 files is included in Relion.RelionParticleData
+
+    def write_file(self, file_name=None, additional_files=None):
+
+        # Get the X, Y, and Z sizes from widget
+        dialog = CoordInputDialog()
```

Comment:
We have to change this a little bit because opening a widget makes it unusable via the command line interface in ChimeraX. There is an existing method for getting additional information from the save dialog.

Reply:
The PEET saver has an example of this, but let me take a stab at modifying.